contracts/vaults/balancer/internal/pool/BalancerComposablePoolUtils.sol

```diff
@@ -0,0 +1,451 @@
+        BalancerComposablePoolContext memory poolContext,
+        StrategyContext memory strategyContext,
+        ComposableRedeemParams memory params,
+        uint256[] memory exitBalances
+    ) internal returns (uint256 primaryPurchased) {
+        address[] memory tokens = poolContext.basePool.tokens;
+        uint256 primaryIndex = poolContext.basePool.primaryIndex;
+        uint256 tradeIndex;
+        for (uint256 i; i < tokens.length; i++) {
+            if (i == poolContext.bptIndex) continue;
+
+            if (i == primaryIndex) {
+                primaryPurchased += exitBalances[i];
+            } else {
+                if (exitBalances[i] > 0) {
+                    primaryPurchased += _sellToken({        
+                        strategyContext: strategyContext,
+                        params: params.redemptionTrades[tradeIndex],
+                        sellToken: tokens[i],
+                        buyToken: tokens[primaryIndex],
+                        sellAmount: exitBalances[i]
+                    });
+                }
+                tradeIndex++;
+            }
+        }
+    }
+
+    function _redeem(
+        BalancerComposablePoolContext memory poolContext,
+        StrategyContext memory strategyContext,
+        AuraStakingContext memory stakingContext,
+        uint256 strategyTokens,
+        ComposableRedeemParams memory params
+    ) internal returns (uint256 finalPrimaryBalance) {
+        uint256 bptClaim = strategyContext._redeemStrategyTokens(strategyTokens);
+        bool isSingleSidedExit = params.redemptionTrades.length == 0;
+
+        // Underlying token balances from exiting the pool
+        uint256[] memory exitBalances = _unstakeAndExitPool(
+            poolContext, stakingContext, bptClaim, params.minAmounts, isSingleSidedExit
+        );
+
+        finalPrimaryBalance = _convertTokensToPrimary(
+            poolContext, strategyContext, params, exitBalances
+        );
+    }
+
+    function _joinPoolAndStake(
+        BalancerComposablePoolContext memory poolContext,
+        ComposableOracleContext memory oracleContext,
+        StrategyContext memory strategyContext,
+        AuraStakingContext memory stakingContext,
+        uint256[] memory amounts,
+        uint256 minBPT
+    ) internal returns (uint256 bptMinted) {
+        // prettier-ignore
+        PoolParams memory poolParams = _getPoolParams({
+            context: poolContext,
+            amounts: amounts,
+            isJoin: true,
+            isSingleSided: false,
+            bptAmount: minBPT
+        });
+
+        bptMinted = BalancerUtils._joinPoolExactTokensIn({
+            poolId: poolContext.poolId,
+            poolToken: poolContext.basePool.poolToken,
+            params: poolParams
+        });
+
+        // Check BPT threshold to make sure our share of the pool is
+        // below maxPoolShare
+        uint256 bptThreshold = strategyContext.vaultSettings._poolClaimThreshold(
```

Comment:
Is this bptThreshold number calculated on the old virtualSupply? I.e. the virtualSupply before we executed our join on line 349.

Reply:
Wouldn't our join affect this threshold?

Reply:
fixed, using getActualSupply() from the pool